### PR TITLE
WIP: Allow overriding WasmResult of mocked foreign functions

### PR DESCRIPTION
**Pull request by jamesmulcahy**
I'd like to be able to generate better coverage of my foreign functions, but right now, the code limits you to an implicit "Ok" result.  This prevents me testing some of the error paths behaviors in my WASM extensions.

This PR is an attempt to make that interface more generic. The problem I see is that I need a "WasmResult" type, but the only existing thing like that is internal.Status.....which is internal.

I've added an explicit WasmResult type, but with obvious downsides (internal.Status and types.WasmResult need to be in sync, otherwise things will get nasty).

I'm not 100% sure on the best path forward, but thought I'd open this PR to drive a discussion.  I welcome feedback on better/alternate approches!

### CODE REVIEW

File: proxywasm/proxytest/root.go [hunk 1]
```diff
@@ -276,12 +276,17 @@ func (r *rootHostEmulator) ProxyCallForeignFunction(funcNamePtr *byte, funcNameS
 	f, ok := r.foreignFunctions[funcName]
 	if !ok {
 		log.Fatalf("%s not registered as a foreign function", funcName)
+		return internal.StatusNotFound
+	}
+
+	ret, retval := f(param)
+
+	if len(ret) > 0 {
```
Comment (jamesmulcahy) [inline proxywasm/proxytest/root.go:284]: I also added a fix for zero byte responses; this looked problematic as is.  Happy to split this to another PR if you like.

**Comment by mathetake** [comment]
sorry @jamesmulcahy I will take a look at this within next 24 hrs..

File: proxywasm/types/types.go [hunk 2]
```diff
@@ -80,3 +80,21 @@ var (
...
+	WasmResultParseFailure         WasmResult = 4
+	WasmResultBadExpression        WasmResult = 5
+	WasmResultInvalidMemoryAccess  WasmResult = 6
+	WasmResultEmpty                WasmResult = 7
+	WasmResultCasMismatch          WasmResult = 8
+	WasmResultResultMismatch       WasmResult = 9
+	WasmResultInternalFailure      WasmResult = 10
+	WasmResultBrokenConnection     WasmResult = 11
+	WasmResultUnimplemented        WasmResult = 12
+)
```
Comment (mathetake) [inline proxywasm/types/types.go:100]: I don't think I could accept this approach, as our `types` and `proxywasm` packages are abstraction layers to hide the ABI detail (e.g. enums, rawhostcall, etc), and I intentionally wraps all the status in internal package as errors for example (see the a few lines above where you can  find all the errors and docs.

File: proxywasm/types/types.go [hunk 3]
```diff
@@ -80,3 +80,21 @@ var (
 	// ErrorUnimplemented indicates the API is not implemented in the host yet.
 	ErrorUnimplemented = errors.New("error status returned by host: unimplemented")
 )
+
+type WasmResult uint32
+
+const (
+	WasmResultOk                   WasmResult = 0
```
Comment (jcchavezs) [inline proxywasm/types/types.go:87]: Is this already imposed somewhere? Otherwise we can use `iota`.
